Src/ILGPU.Algorithms.Tests/MatrixTests.cs

```
@@ -0,0 +1,258 @@
+        private static void AssertMatrixEqual2D(
+            float[,] left,
+            float[,] right,
+            float eps = 1e-4f)
+        {
+            var leftRows = left.GetLength(0);
+            var leftColumns = left.GetLength(1);
+            var rightRows = right.GetLength(0);
+            var rightColumns = right.GetLength(1);
+
+            if (leftRows != rightRows || leftColumns != rightColumns)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(right),
+                    $"Matrix dimensions {leftRows}x{leftColumns} and " +
+                    $"{rightRows}x{rightColumns} do not match");
+            }
+
+            for (var i = 0; i < leftRows; i++)
+            {
+                for (var j = 0; j < leftColumns; j++)
+                {
+                    Assert.True(
+                        Math.Abs(left[i, j] - right[i, j]) < eps,
+                        "Matrix result not equal");
+                }
+            }
+        }
+
+        #endregion
+
+        [SkippableTheory()]
+        [MemberData(nameof(DimensionsData))]
+        public void MatrixMultiplySparse(
+            int length,
+            int baseLength,
+            int seed,
+            float sparsityProbability)
+        {
+            Skip.If(length > 512);
+            
+            using var stream = Accelerator.CreateStream();
+            
+            // Our starting dense matrix
+            var denseMatrix = new float[baseLength, length];
+
+            // Setup sparse 2D matrix
+            var sparseMatrix = new float[length, length];
+            var sparseMaskMatrix = new float[length, length];
+
+
+            var random = new System.Random(seed);
+            for (int i = 0; i < denseMatrix.GetLength(0); ++i)
+            {
+                for (int j = 0; j < denseMatrix.GetLength(1); ++j)
+                {
+                    denseMatrix[i, j] = random.NextSingle();
```

I don't suppose there is a way to suppress this error about NextSingle? The page of warnings from Github makes this hard to read.